dev-tools/releng/src/main.rs

```diff
@@ -0,0 +1,734 @@
+use slog::error;
+use slog::info;
+use slog::Drain;
+use slog::Logger;
+use slog_term::FullFormat;
+use slog_term::TermDecorator;
+use tokio::sync::Semaphore;
+
+use crate::cmd::Command;
+use crate::job::Jobs;
+
+/// The base version we're currently building. Build information is appended to
+/// this later on.
+///
+/// Under current policy, each new release is a major version bump, and
+/// generally referred to only by the major version (e.g. 8.0.0 is referred
+/// to as "v8", "version 8", or "release 8" to customers). The use of semantic
+/// versioning is mostly to hedge for perhaps wanting something more granular in
+/// the future.
+const BASE_VERSION: Version = Version::new(8, 0, 0);
+
+#[derive(Debug, Clone, Copy)]
+enum InstallMethod {
+    /// Unpack the tarball to `/opt/oxide/<service-name>`, and install
+    /// `pkg/manifest.xml` (if it exists) to
+    /// `/lib/svc/manifest/site/<service-name>.xml`.
+    Install,
+    /// Copy the tarball to `/opt/oxide/<service-name>.tar.gz`.
+    Bundle,
+}
+
+/// Packages to install or bundle in the host OS image.
+const HOST_IMAGE_PACKAGES: [(&str, InstallMethod); 7] = [
+    ("mg-ddm-gz", InstallMethod::Install),
+    ("omicron-sled-agent", InstallMethod::Install),
+    ("overlay", InstallMethod::Bundle),
+    ("oxlog", InstallMethod::Install),
+    ("propolis-server", InstallMethod::Bundle),
+    ("pumpkind-gz", InstallMethod::Install),
+    ("switch-asic", InstallMethod::Bundle),
+];
+/// Packages to install or bundle in the recovery (trampoline) OS image.
+const RECOVERY_IMAGE_PACKAGES: [(&str, InstallMethod); 2] = [
+    ("installinator", InstallMethod::Install),
+    ("mg-ddm-gz", InstallMethod::Install),
+];
+/// Packages to ship with the TUF repo.
+const TUF_PACKAGES: [&str; 11] = [
+    "clickhouse_keeper",
+    "clickhouse",
+    "cockroachdb",
+    "crucible-pantry-zone",
+    "crucible-zone",
+    "external-dns",
+    "internal-dns",
+    "nexus",
+    "ntp",
+    "oximeter",
+    "probe",
+];
+
+const HELIOS_REPO: &str = "https://pkg.oxide.computer/helios/2/dev/";
+
+static WORKSPACE_DIR: Lazy<Utf8PathBuf> = Lazy::new(|| {
+    // $CARGO_MANIFEST_DIR is at `.../omicron/dev-tools/releng`
+    let mut dir =
+        Utf8PathBuf::from(std::env::var("CARGO_MANIFEST_DIR").expect(
+            "$CARGO_MANIFEST_DIR is not set; run this via `cargo xtask releng`",
+        ));
+    dir.pop();
+    dir.pop();
+    dir
+});
+
+#[derive(Parser)]
+/// Run the Oxide release engineering process and produce a TUF repo that can be
+/// used to update a rack.
+///
+/// For more information, see `docs/releng.adoc` in the Omicron repository.
+///
+/// Note that `--host-dataset` and `--recovery-dataset` must be set to different
+/// values to build the two OS images in parallel. This is strongly recommended.
+struct Args {
+    /// ZFS dataset to use for `helios-build` when building the host image
+    #[clap(long, default_value_t = Self::default_dataset("host"))]
+    host_dataset: String,
```

Comment:
nit: is there some validation we can do on the dataset names? Would that be worth it at all?

Reply:
I think verifying the datasets already exist (one of the preflight checks we run) is sufficient.